platform_plugin_forum_email_notifier/tasks.py

```diff
@@ -85,25 +134,150 @@ def notify_users(thread_id, discussion, course_id, body, title, url, object_type
     subscribers = subscribers | staff_subscribers
 
     for subscriber in subscribers:
-        try:
-            user = User.objects.get(id=subscriber)
-            preference = ForumNotificationPreference.objects.get(
-                user=user, course_id=course_id
-            )
-            if preference.preference in (
-                PreferenceOptions.NONE,
-                PreferenceOptions.ALL_POSTS_DAILY_DIGEST,
-                PreferenceOptions.ALL_POSTS_WEEKLY_DIGEST,
-            ):
+        if subscriber in staff_subscribers:
+            try:
+                user = User.objects.get(id=subscriber)
+                # Altough we are getting staff_subscribers whom have a preference of ALL_POSTS,
+                # they can already be a subscriber of the thread, so we need to check if they
+                # have a preference for the course
+                preference = ForumNotificationPreference.objects.get(
+                    user=user, course_id=course_id
+                )
+                if preference.preference in (
+                    PreferenceOptions.NONE,
+                    PreferenceOptions.ALL_POSTS_DAILY_DIGEST,
+                    PreferenceOptions.ALL_POSTS_WEEKLY_DIGEST,
+                ):
+                    continue
```

Comment:
how can I have multiple preferences?

Reply:
Does this mean that daily, weekly digests have precedence over all posts?

Reply:
Not, daily, weekly, and none just have a different behavior than ALL_POST. All post is notified immediately while other are digested and some prefer to not receive any notification.

That's why we are skipping all of that. And digests are handled separately

Reply:
You can't have multiple preferences for the same course id and the same user. There is a database constraint for that.

Reply:
That's why I asked. Aren't staff subscribers already filtered by all posts?

Reply:
[curious] My question is: if there's a constraint on user, course id, how can [this](https://github.com/eduNEXT/platform-plugin-forum-email-notifier/pull/4/files#diff-d6f8aa8c3a53ed681d96a188fd78330e8a56fafecdc0171b67c325cccb2648adR140-R142) happen?

Reply:
yes, but they can a subcribers of the thread in the forum, and that record is different from the custom one in `ForumPreference`. That's why we are filtering staff subcribers.

it seems we can refactor this one, but I'm not sure of the right direction